.github/workflows/build-plugin-archive.yml

```diff
@@ -118,82 +109,174 @@ jobs:
 
-      - name: Check for package.json
-        id: check-for-package
+      - name: Set up node cache mode
         run: |
-          if [ -f package.json ]; then
-            echo "has_package=true" >> $GITHUB_OUTPUT
+          if [ "${{ inputs.PACKAGE_MANAGER }}" == 'npm' ] && { [ -f "${GITHUB_WORKSPACE}/package-lock.json" ] || [ -f "${GITHUB_WORKSPACE}/npm-shrinkwrap.json" ]; }; then
+            echo "NODE_CACHE_MODE=npm" >> $GITHUB_ENV
+          elif [ "${{ inputs.PACKAGE_MANAGER }}" == 'yarn' ] && [ -f "${GITHUB_WORKSPACE}/yarn.lock" ]; then
+            echo "NODE_CACHE_MODE=yarn" >> $GITHUB_ENV
           else
-            echo "has_package=false" >> $GITHUB_OUTPUT
+            echo "No lock files found or unknown package manager"
           fi
 
       - name: Set up node
-        if: steps.check-for-package.outputs.has_package == 'true'
+        if: steps.composer-tools.outputs.assets-compiler == '0'
         uses: actions/setup-node@v4
         with:
           node-version: ${{ inputs.NODE_VERSION }}
           registry-url: ${{ inputs.NPM_REGISTRY_DOMAIN }}
           cache: ${{ env.NODE_CACHE_MODE }}
 
-      - name: Run Composer Asset Compiler
-        if: steps.composer-tools.outputs.assets-compiler == '0' && steps.check-for-package.outputs.has_package == 'true'
-        run: composer compile-assets ${{ inputs.COMPILE_ASSETS_ARGS }}
+      - name: Install and run Composer Asset Compiler
+        if: steps.composer-tools.outputs.assets-compiler == '0'
+        run: |
+          composer global config --no-plugins --no-interaction allow-plugins.inpsyde/composer-assets-compiler true
+          composer global require inpsyde/composer-assets-compiler
+          composer compile-assets ${{ inputs.COMPILE_ASSETS_ARGS }}
 
-      - name: Run WordPress Translation Downloader
+      - name: Install and run WordPress Translation Downloader
         if: steps.composer-tools.outputs.translation-downloader == '0'
-        run: composer wp-translation-downloader:download
+        run: |
+          composer global config --no-plugins --no-interaction allow-plugins.composer/installers true
+          composer global config --no-plugins --no-interaction allow-plugins.inpsyde/wp-translation-downloader true
+          composer global require inpsyde/wp-translation-downloader
+          composer wp-translation-downloader:download
 
-      - name: Install Composer dependencies without dev dependencies
-        # ⬑ Now we want to get rid of dev dependencies and most probably skip composer scripts
-        uses: ramsey/composer-install@v3
+      - name: Run PHP Scoper
+        if: steps.composer-tools.outputs.php-scoper == '0'
+        run: |
+          php-scoper add-prefix --force --output-dir build
+          composer --working-dir=build dump-autoload -o
+          sed -i "s/'__composer_autoload_files'/\'__composer_autoload_files_${{ github.sha }}'/g" "build/vendor/composer/autoload_real.php"
```

Comment:
This one warrants some attention. When autoloading files, composer checks a hash of `<PACKAGE>+<PATH>` against a list in `$GLOBALs['__composer_autoload_files']`. If the hash is already present, the file will not be included.

Since composer does not care about the **file contents** in this hash, it will prevent the prefixed version of the file from being included (despite clearly being a different piece of code now). This essentially breaks "Inpsyde Assets" because it relies on autoloaded files, and -when prefixed- fails to instantiate if another instance is present on the WP installation.

Since the intent is clearly to implement "full prefixing" of the entire plugin (as opposed to scoping away only a few known incompatibilities), we do not expect major issues from potentially allowing duplicate includes through this workaround. Yet we obviously cannot rule them out either. I checked with colleagues and we're willing to dogfood this a little and accept that risk.

Reply:
While I'm okay with including this step, it is fragile. More importantly, I wonder whether this should not be a feature of PHP-Scoper instead. Does it make sense to open an issue there and try to get this "fixed" upstream?

Reply:
https://github.com/humbug/php-scoper/issues/298
https://github.com/humbug/php-scoper/issues/1036

[There also used to be documentation about this](https://github.com/humbug/php-scoper/pull/1046/files), but it has been removed a while ago. Maybe this is worth another look..

Reply:
I checked this again: This isn't actually an issue with php-scoper - it prefixes the files just fine. The problem is that *composer* refuses to load "the same file twice" and it decides that based on the relative file path, ignoring the actual file contents (which are different and non-conflicting now after scoping).

IIRC we tried to implement post-processing to salt the hashes via scope-config, but weren't able to come up with a universal and easy solution. So doing it in the GHA appeared like the logical next alternative.

@tyrann0us